enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,421 @@
+internal loadbalancer implementation has been around for a number of
+releases at this point and we are just extending it.
+
+#### Dev Preview -> Tech Preview
+
+NA
+
+#### Tech Preview -> GA
+
+NA
+
+#### Removing a deprecated feature
+
+NA
+
+### Upgrade / Downgrade Strategy
+
+Upgrades and downgrades will be handled the same way they are for the current
+internal loadbalancer implementation. On upgrade, existing VIP configurations
+will be maintained. We will not (and cannot) automatically add new VIPs to a
+cluster on upgrade. If a deployer of an older dual stack cluster wants the new
+VIP functionality that will have to be a separate operation from upgrade.
+
+### Version Skew Strategy
+
+This will also be handled as it is today. The keepalived configuration used
+for the new version must work with the old version until all nodes have been
+upgraded. If breaking changes are needed, there will be a migration mechanism
+that runs after upgrade is complete (similar to what we did in the multicast
+to unicast migration).
+
+### Operational Aspects of API Extensions
+
+NA
+
+#### Failure Modes
+
+NA
+
+#### Support Procedures
+
+NA
+
+## Implementation History
+
+4.11: Initial implementation
+
+## Drawbacks
+
+Nothing significant. A very small amount of compute resources will be used
+to manage the new VIPs. Even this can be avoided by simply not specifying
+the second VIP if it is not needed.
+
+## Alternatives
+
+Implement this somewhat like the
+[existing workaround](https://github.com/yboaron/kepalived-v6) - create a
+separate instance of keepalived that is responsible for ipv6 only. This
+results in some duplication of containers and configs, but it would probably
+simplify the code changes.
+
+### Alternative install-config layouts
+
+#### Just add v6 versions of the VIP fields
+```yaml
+platform:
+  baremetal:
+    apiVIP: 192.168.111.5
+    apiVIPv6: f00::5
+    ingressVIP: 192.168.111.4
+    ingressVIPv6: f00::4
+```
+This is a minimal change, but it leaves us with somewhat of a problem in that
+we can't assume the old fields always contain a v4 address since previous
+releases allowed either v4 or v6. We may be able to deal with that by
+detecting the appropriate ip version for the old fields and populating new
+fields on the api side. E.g. if apiVIP is v4, we populate
+`platformStatus.apiServerInternalIPv4` and if it is v6 we populate
+`apiServerInternalIPv6` instead. That way the only place we need to worry about
+the difference is in the installer. We would need a validation to ensure a v6
+VIP isn't specified in both fields.
+
+Note that this assumes we replace the existing api fields with two new ones
+per VIP and do not reuse the existing ones at all. I think we're going to want
+to do that regardless of what install-config layout we use so we don't need
+logic on the backend in machine-config-operator and baremetal-runtimecfg to
+handle older cluster configs.
+
+#### Deprecate the old fields and add new v4 and v6 fields
```

Comment:
This would be my pick for the installer API change. The fields are unambiguous to the end user. Granted that this option is a little more intensive to implement, it shouldn't be too complicated. We can start using the new v4/v6 fields right away and add the correct default (for ipv4, ipv4 or dualstack) to the deprecated fields.

Reply:
This was the frontrunner initially, but Dan pointed out that if we're going to deprecate the old fields it would be more consistent with the other network settings to make the new fields lists (as in the next alternative).

That said, I'm not opposed to any of these options (except the one you noted above), so if the installer team prefers this I'm not inclined to argue too strenuously. :-)

Reply:
Most unambiguous. Again we need a way for the user to indicate what the primary VIP is if that is even required.